client/extensions/woocommerce/app/dashboard/index.js

```diff
@@ -50,49 +50,61 @@ class Dashboard extends Component {
 		className: PropTypes.string,
 		finishedInitialSetup: PropTypes.bool,
 		hasOrders: PropTypes.bool,
+		isSetupComplete: PropTypes.bool,
 		loading: PropTypes.bool,
 		selectedSite: PropTypes.shape( {
 			ID: PropTypes.number.isRequired,
 			slug: PropTypes.string.isRequired,
 			URL: PropTypes.string.isRequired,
 		} ),
+		siteId: PropTypes.number,
 		mailChimpConfigured: PropTypes.bool,
 		fetchOrders: PropTypes.func,
 		fetchSetupChoices: PropTypes.func,
 		requestSyncStatus: PropTypes.func,
+		setupChoicesLoading: PropTypes.bool,
 	};
 
 	componentDidMount = () => {
-		const { selectedSite, productsLoaded } = this.props;
+		const { siteId, productsLoaded } = this.props;
 
-		if ( selectedSite && selectedSite.ID ) {
-			this.props.fetchSetupChoices( selectedSite.ID );
-			this.props.fetchOrders( selectedSite.ID );
-			this.props.requestSettings( selectedSite.ID );
+		if ( siteId ) {
+			this.props.fetchSetupChoices( siteId );
+			this.props.fetchOrders( siteId );
+			this.props.requestSettings( siteId );
 
 			if ( ! productsLoaded ) {
-				this.props.fetchProducts( selectedSite.ID, { page: 1 } );
+				this.fetchSiteProducts( siteId );
 			}
 		}
 	};
 
 	componentWillReceiveProps = newProps => {
-		const { selectedSite, productsLoaded } = this.props;
+		const { siteId, productsLoaded } = this.props;
 
-		const newSiteId = newProps.selectedSite ? newProps.selectedSite.ID : null;
-		const oldSiteId = selectedSite ? selectedSite.ID : null;
+		const newSiteId = newProps.siteId ? newProps.siteId : null;
+		const oldSiteId = siteId ? siteId : null;
 
 		if ( newSiteId && oldSiteId !== newSiteId ) {
 			this.props.fetchSetupChoices( newSiteId );
 			this.props.fetchOrders( newSiteId );
 			this.props.requestSettings( newSiteId );
 
 			if ( ! productsLoaded ) {
-				this.props.fetchProducts( newSiteId, { page: 1 } );
+				this.fetchSiteProducts( newSiteId );
 			}
 		}
 	};
 
+	fetchSiteProducts = siteId => {
+		const params = { page: 1 };
+		const failureAction = dispatch => {
+			dispatch( this.props.fetchProductsFailure( siteId, params ) );
+		};
+
+		this.props.fetchProducts( siteId, params, null, failureAction );
```

Comment:
We want to set the `isLoading` state flag to `false` every time the `fetchProducts` request fails, no matter where it's called from, don't we? In that case the `failureAction` should be dispatched inside the data layer, not in app code.

The best place to dispatch this is the `apiError` function in `state/data-layer/products/index.js`.

Setting the `isLoading` when the request is initiated and when it succeeds is also performed inside data layer. The caller of `fetchProducts` doesn't need to have the responsibility for doing that.

By the way, the `fetchProducts` action creator also has a `successAction` parameter, but it's never used, even when I pass a non-null value. That's an unrelated bug report for @coderkevin

Reply:
Indeed, I agree - and had actually started putting the logic there. I will refactor this a bit and add some tests too.